README
# Sticky

Sticky is a simple, key/value pair, browser-storage cache leveraging the latest HTML5 storage API's (indexedDB, webSQL, and localStorage).

Sticky persists to memory, indexedDB, webSQL, localStorage, globalStorage, and cookies.

Objects and arrays are stringified before storage and strings longer than 128 characters aren't persisted to cookies.

## Browser Support

* **WebSQL (SQLite)**  
Chrome 4+, Opera 10.5+, Safari 3.1+, and Android Browser 2.1+
5MB of data per DB, but can request more
* **IndexedDB**  
IE 10+, FireFox 4+, and Chrome 11+
* **localStorage**  
Safari 4+, Mobile Safari (iPhone/iPad), Firefox 3.5+, Internet Explorer 8+ and Chrome 4+
5MB of data per domain
* **globalStorage**  
FireFox 2-3
* **Cookies**  
Maximum size varies across implementations

For more compatibility information, see: [http://caniuseit.com/](http://caniuseit.com/)

## Usage

#### HTML:

    <script src="sticky-0.9.js" type="text/javascript"></script>

#### JavaScript:

    // Initialize your store and repopulate cached data
    var store = new Sticky();

    // Set
    store.set('color', 'red');

    // Get
    store.get('color');

    // Remove
    store.remove('color');

### Initialize a store

First, you must create a sticky store object like so:

    var store = new Sticky();

Alternatively, you can specify some options for this store by passing the opts argument:

    var store = new Sticky({
        domain: 'example.com', // Used for cookie domain.
        expires: 48,           // Hours. Used for cookie expiration.
        ready: function(){},   // Fires after cache has been repopulated.
        size: 5,               // indexedDB / webSQL database size in megabytes.
        version: '1.0'         // indexedDB / webSQL database version.
    });

When you initialize a store, it's cache will be repopulated from browser storage. Because indexedDB and webSQL operate asynchronously, Sticky will fire the ```store.opts.ready``` function after the cache has been repopulated.


### Get

Returns the cached value or null if it isn't found. Example:

    var store = new Sticky();

    store.get('something');

### Set

Caches a value and returns the cached value or false on error. You can pass any type of value: String, array, object, and number. Objects and arrays will be stringified and prefixed with ```J::O``` for storage.

    var store = new Sticky();

    store.set('car', {
        make: 'Volkswagen',
        model: 'Golf GTI',
        year: 2001
    });

### Remove

Removes the cached value from all storage mechanisms and returns true if successful.

    var store = new Sticky();

    store.remove('something');

## License

Copyright 2011 Alexander C. Mingoia

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
# Release History

# 2.11 / 2014-06-22

* Publish to bower package registry.

# 2.10 / 2014-05-27

* Allow falsy values to be saved (empty string and number 0).
* Fix expect.js URL in browser tests.

# 2.9 / 2013-03-03

* Fix issue with updated indexedDB spec: IDBTransaction now has mode argument
  which takes a string, instead of the previously used prototype properties.

# 2.8 / 2012-03-08

* Fix issue where indexedDB adapter's get method didn't execute callback when lookup is successful but result is empty.

# 2.7 / 2012-03-07

* Improved get and set methods by handling array of keys/items.
* Removed redundant serialization in indexedDB adapter's get and set
  methods.

## 2.6 / 2012-02-23

* Ensure compatibility with both FF and Chrome indexedDB implementations
* Added tests for each adapter (using Mocha framework)
* Refactored cookie adapter methods and fixed bug with StickyStore.remove() being called twice in certain cases
* Fixed bug with missing callback in StickyStore.adapters.localStorage.remove() causing multiple removals not to fire callback
* Fixed bug with number type conversion in StickyStore.serialize()

## 2.5 / 2012-02-19

* Fixed bug with callback not firing after all storage adapters are exhausted

## 2.4 / 2012-02-18

* Cleaned-up StickyStore.exec()

## 2.3 / 2012-02-18

* Fixed bug with indexedDB initialization not firing callback in certain
  cases.

## 2.2 / 2012-02-15

* Fixed localStorage adapter's get method not unserializing

## 2.1 / 2012-02-13

* Instead of waiting for all storage interfaces to connect before firing
  ready event, only wait for one interface to connect.

## 2.0 / 2012-02-13

* Added support for new IDBRequest.onupdateneeded event, while keeping
  backwards-compatible with old IDBDatabase.setVersion() method.
* Re-written for modularization of storage mechanisms.
* Proper support for multiple stores.

## 1.2 / 2011-12-29

* Fixed typo in webSQL removal callback

## 1.1 / 2011-12-23

* Added events; StickyStore.on() and StickyStore.trigger()
* Aliased the ready option by registering it with the 'ready' event
* Expanded error handling and replaced console messages with 'error' events
* Added events for StickyStore.get(), StickyStore.set(), and StickyStore.remove()

## 1.0 / 2011-12-22

* Improved error handling

## 0.9 / 2011-12-16

* Convert stored numbers to Number objects when repopulating

## 0.8 / 2011-11-17

* Fixed bug with storing Number objects
* Refactored some indexedDB logic when repopulating the store
* Refactored type checking in StickyStore.set()

## 0.7 / 2011-11-01

* Removed versioning of stores due to too many idiosyncracies
* Fixed bug with repopulating previous sessions data using indexedDB

## 0.6 / 2011-10-24

* Fixed bug with key reference for ObjectStore.put()

## 0.5 / 2011-10-21

* Fixed bug with webSQL versions
* Fixed bug with storing objects in indexedDB
* Updated documentation

## 0.4 / 2011-10-16

* Added callbacks
* Added default value for store.get()
* Added browser-specific detection of indexedDB

## 0.3 / 2011-10-11

* Fixed bug with indexedDB's continue and delete methods being reserved keywords causing Safari to throw errors

## 0.2 / 2011-10-10

* Added ```store.opts.name``` to handle multiple stores by prefixing keys
* Added ```store.removeAll()``` method
* Fixed bug where only objects and arrays were persisted to webSQL and indexedDB
* Updated documentation

## 0.1 / 2011-10-09

* Initial release
* Started release history


sticky-0.9.js
/**
 * Sticky
 *
 * Version 0.9
 * Copyright 2011 Alexander C. Mingoia
 * MIT Licensed
 *
 * Simple JavaScript HTML5 browser storage cache. Persists to
 * memory, indexedDB, webSQL, localStorage, globalStorage, and cookies.
 *
 * Objects and arrays are stringified before storage, and
 * strings longer than 128 characters aren't persisted to cookies.
 *
 * WebSQL (SQLite)
 *     Chrome 4+, Opera 10.5+, Safari 3.1+, and Android Browser 2.1+
 *     5MB of data per DB, but can request more
 * IndexedDB
 *     IE 10+, FireFox 4+, and Chrome 11+
 * localStorage
 *     Safari 4+, Mobile Safari (iPhone/iPad), Firefox 3.5+, Internet Explorer 8+ and Chrome 4+
 *     5MB of data per domain
 * globalStorage
 *     FireFox 2-3
 * Cookies
 *     1-4KB depending on character encoding and implementation.
 *
 * For more compatibility information, see: http://caniuse.com/
 */


/**
 * Constructor
 *
 * @param {Map} opts Options:
 *    var options = {
 *      name: 'Store A'          // A name for this store that acts as a unique identifier.
 *                               // You need to set this if you're going to have multiple stores.
 *      domain: 'example.com',   // Custom cookie domain
 *      expires: 168,            // Cookie expiration in hours
 *      ready: function() {},    // Fires after repopulating cache
 *      size: 10                 // WebSQL database size in megabytes
 *    };
 *
 *  @return {StickyStore} Returns an instantiated store object
 */

function StickyStore(opts) {
  // Default options
  if (!opts) opts = {};
  if (!opts.domain) opts.domain = window.location.hostname;
  if (!opts.expires) opts.expires = 24*7; // Cookie expiration in hours
  if (!opts.name) opts.name = 'sticky';
  if (!opts.size) opts.size = 5; // Size in MB
  if (opts.ready && typeof opts.ready !== 'function') {
      throw new Error('opts.ready callback must be a function');
  }

  this.opts = opts;
  this.cache = {}; // Memory cache container object
  this.db; // Indexed DB or Web SQL connection object

  // Needed to keep context in async methods
  var store = this;

  // Wrap localStorage and globalStorage
  if (window.localStorage) {
    this.storage = localStorage;
  }
  else if (window.globalStorage) {
    this.storage = globalStorage[opts.domain];
  }

  // Load items from localStorage
  if (this.storage) {
    for (var i=0; i<this.storage.length; i++) {
      var record = this.storage.key(i);
      var data = this.storage.getItem(key);
      if (data && record.indexOf(opts.name) === 0) {
        var key = record.replace(new RegExp(opts.name.replace(/[^\w]/gi, ''), 'gi'), '');
        var item;
        // Object
        if (data.indexOf('J::O') === 0) {
          try {
            item = JSON.parse(data.slice(4));
          }
          catch (err) {
            console.log('Sticky Error: ' + err);
          }
        }
        // String
        else if (isNaN(data)) {
            item = data;
        }
        // Number
        else {
            item = Number(data);
        }
        if (item) {
          store.set(key, item);
        }
      }
    }
  }

  // Initialize IndexedDB and repopulate cache
  if ('mozIndexedDB' in window) {
     window.indexedDB = window.mozIndexedDB;
  }
  if (window.indexedDB) {
    // Request DB
    var request = window.indexedDB.open(opts.name+'_sticky', 'Sticky Offline Web Cache');
    request.onsuccess = function(event) {
      store.db = event.target.result
      // If version is different, we need to set version
      // and create an object store
      if (store.db.version != '0.8') {
        var request = store.db.setVersion('0.8');
        request.onsuccess = function(event) {
          // Create our object store for cached data
          if (!store.db.objectStoreNames.contains('cache')) {
              store.db.createObjectStore('cache', {keyPath: 'key'});
          }
          opts.ready && opts.ready.call(store);
        };
        request.onerror = function(event) {
          console.log('Sticky Error: ' + request.errorCode);
          opts.ready && opts.ready.call(store);
        };
      }
      else {
        var transaction = store.db.transaction('cache');
        var objectStore = transaction.objectStore('cache');
        objectStore.openCursor().onsuccess = function(event) {
          var cursor = event.target.result;
          // Only load records for this specific store
          if (cursor && cursor.value.data && cursor.key.indexOf(opts.name) === 0) {
            var key = cursor.key.replace(new RegExp(opts.name.replace(/[^\w]/gi, ''), 'gi'), '');
            var item;
            // Object
            if (cursor.value.data.indexOf('J::O') === 0) {
              try {
                item = JSON.parse(cursor.value.data.slice(4));
              }
              catch (err) {
                console.log('Sticky Error: ' + err);
              }
            }
            // String
            else if (isNaN(cursor.value.data)) {
                item = cursor.value.data;
            }
            // Number
            else {
                item = Number(cursor.value.data);
            }
            if (item) {
              store.set(key, item);
            }
            cursor['continue']();
          }
          else {
            opts.ready && opts.ready.call(store);
          }
        }
      }
    }
    request.onerror = function(event) {
      console.log("Sticky Error: Couldn't open Indexed DB (Code " + request.errorCode + ")");
      opts.ready && opts.ready.call(store);
    }
  }
  // Initialize WebDB and repopulate cache
  else if (window.openDatabase) {
    // Try and open DB
    try {
      this.db = window.openDatabase(opts.name+'_sticky', '0.8', 'Sticky Offline Web Cache', (opts.size * 1024 * 1024));
      if (this.db) {
        this.db.transaction(function(tx) {
          tx.executeSql('CREATE TABLE IF NOT EXISTS cache (key TEXT, data TEXT)');
          // Repopulate cache container object with data stored in SQLite
          tx.executeSql('SELECT * FROM cache', [], function(tx, results) {
            if (results.rows.length > 0) {
              for (var i=0; i<results.rows.length; i++) {
                var record = results.rows.item(i);
                // Only load records for this specific store
                if (record['data'] && record['key'].indexOf(opts.name) === 0) {
                  var key = record['key'].replace(new RegExp(opts.name.replace(/[^\w]/gi, ''), 'gi'), '');
                  var item;
                  // Object
                  if (record['data'].indexOf('J::O') === 0) {
                    try {
                      item = JSON.parse(record['data'].slice(4));
                    }
                    catch (err) {
                      console.log('Sticky Error: ' + err);
                    }
                  }
                  // String
                  else if (isNaN(record['data'])) {
                      item = record['data'];
                  }
                  // Number
                  else {
                      item = Number(record['data']);
                  }
                  if (item) {
                    store.set(key, item);
                  }
                }
              }
              opts.ready && opts.ready.call(store);
            }
            else {
              opts.ready && opts.ready.call(store);
            }
          });
        });
      }
    }
    catch (err) {
      console.log('Sticky Warning: ' + err);
      opts.ready && opts.ready.call(store);
    }
  }
  else {
    opts.ready && opts.ready.call(store);
  }
};


/**
 * Set
 *
 * @param String key
 * @param Mixed value
 * @param Function callback Optional. Called after async operations are completed.
 *
 * @return Mixed Returns reference to stored value or false for failure or error
 */

StickyStore.prototype.set = (function(key, item, callback) {
  if (!item) return false;
  if (callback && typeof callback !== 'function') {
    throw new Error('Callback must be a function');
    return false;
  }

  // Prefix key with store name/identifier
  var key = (this.opts.name + key).replace(/[^\w]/gi, '');

  var value = item;
  var itemType = typeof value;

  // Store item in memory cache
  this.cache[key] = item;

  // Objects and arrays are stringified, and aren't stored in cookies (they're too big)
  if (itemType !== 'string' ) {
    if (itemType === 'object' || itemType === 'array') {
      try {
        value = 'J::O' + JSON.stringify(value);
      }
      catch (err) {
        console.log('Sticky Error: ' + err);
      }
    }
    else {
        value = value.toString();
    }
  }

  if (value) {
    // Only string values less than 128 characters get stored in cookies
    if (value.length < 128) {
      document.cookie = key + '=' + value
        + '; expires=' + new Date(new Date().getTime() + (this.opts.expires*60*60*1000)).toGMTString()
        + '; path=/';
    }
    // Copy value to localStorage or globalStorage
    if (this.storage) {
      try {
        this.storage.setItem(key, value);
      }
      catch (err) {
        console.log('Sticky Error: ' + err);
      }
    }

    if (this.db) {
      var store = this; // Context for callback
      if (this.db.setVersion) {
        // Copy value to indexedDB
        var tx = this.db.transaction(['cache'], IDBTransaction.READ_WRITE, 0);
        var objStore = tx.objectStore('cache');
        var request = objStore.put({'key':key, 'data':value});
        request.onsuccess = function(e) {
          callback && callback.call(store, item);
        };
        request.onerror = function(e) {
          console.log('Sticky Error: ' + e.target.errorCode);
          callback && callback.call(store, false);
        };
      }
      // Copy value to webDB
      else {
        // Insert callback after update
        var insert = function(tx, result) {
          // If update failed then insert
          if (result && result.rowsAffected === 0) {
            tx.executeSql('INSERT INTO cache (key, data) VALUES (?, ?)', [key, value], function(tx, result) {
              if (result && result.rowsAffected === 0) {
                  callback && callback.call(store, false);
              }
              else {
                  callback && callback.call(store, item);
              }
            });
          }
          else {
            callback && callback.call(store, item);
          }
        }
        // Update, and pass insert as callback
        var update = function(tx) {
          tx.executeSql('UPDATE cache SET data=? WHERE key=?', [value, key], insert);
        }
        this.db.transaction(update);
      }
    }
    else {
      callback && callback.call(this, item);
    }

    return this.cache[key];
  }
  return false;
});


/**
 * Get
 *
 * @param String key
 * @param Mixed default/callback Optional. Anything but a function will be used as the default return falue.
 *
 * @return Mixed Returns reference to stored value or false for failure or error
 */

StickyStore.prototype.get = (function(key, callback) {
  var _default = null;

  // Prefix key with store name
  var key = (this.opts.name + key).replace(/[^\w]/gi, '');

  // Not in memory, let's check elsewhere
  if (!this.cache[key]) {
    // Check localStorage or globalStorage
    if (this.storage) {
      var value = this.storage.getItem(key);
      if (value && value.substr(0, 4) === 'J::O') {
        value = JSON.parse(value.substr(4));
      }
      this.cache[key] = value;
    }
    // If not, check cookies
    else {
      var keyEquals = key + "=";
      var cookieArray = document.cookie.split(';');
      for (var i=0; i<cookieArray.length; i++) {
        var cookie = cookieArray[i];
        while (cookie.charAt(0) === ' ') {
          cookie = cookie.substring(1, cookie.length);
        }
        if (cookie.indexOf(keyEquals) === 0) {
          this.cache[key] = cookie.substring(keyEquals.length, cookie.length);
        }
      }
    }
  }
  // Callback
  if (callback) {
    if (typeof callback === 'function') {
      callback.call(this, this.cache[key]);
    }
    else {
      _default = callback;
    }
  }
  // Data inside webDB is loaded into the store on instantiation,
  // so it's available here.
  if (this.cache[key]) {
    return this.cache[key];
  }
  // Not found
  return _default;
});


/**
 * Remove
 *
 * @param String key
 * @param Function callback Optional. Executed after async operations are complete.
 *
 * @returns Bolean
 */

StickyStore.prototype.remove = (function(key, callback) {
  if (callback && typeof callback !== 'function') {
    throw new Error('Callback must be a function');
    return false;
  }

  // Prefix key with store name
  key = (this.opts.name + key).replace(/[^\w]/gi, '');

  // Remove from memory
  if (this.cache[key]) {
    delete this.cache[key];
  }

  // Remove cookie
  document.cookie = key + '=; expires=-1; path=/';

  // Remove localStorage or globalStorage
  if (this.storage) {
    try {
      this.storage.removeItem(key);
    }
    catch (err) {
      console.log('Sticky Error: ' + err);
      return false;
    }
  }

  if (this.db) {
    // Keep context for callback
    var store = this;
    // Remove indexedDB
    if (this.db.setVersion) {
      // Copy value to indexedDB
      var tx = this.db.transaction(['cache'], IDBTransaction.READ_WRITE, 0);
      var objStore = tx.objectStore('cache');
      var request = objStore['delete'](key);
      request.onsuccess = function(e) {
        callback && callback.call(store, true);
      };
      request.onerror = function(e) {
        callback && callback.call(store, false);
      };
    }
    // Remove web SQL
    else {
      this.db.transaction(function(tx) {
        tx.executeSql('DELETE FROM cache WHERE key=?', [key], function(tx, rs) {;
          callback && callback.call(store);
        });
      });
    }
  }
  else {
    callback && callback.call(this);
  }

  return true;
});


/**
 * Remove All
 *
 * @param Function callback Optional
 *
 * Removes all values in this store from all storage mechanisms
 */

StickyStore.prototype.removeAll = (function(callback) {
  var store = this;
  var count = 0;

  var removed = function() {
    count--;
    if (count < 1 && callback && typeof callback === 'function') {
      callback.call(this, this.cache[key]);
    }
  };

  for (var key in store.cache) {
    count++;
  }

  for (var key in store.cache) {
    store.remove(key.replace(this.opts.name, ''), removed);
  }
});
